### PR TITLE
updated inner_call for hashing fn to follow generic approach

### PR DESCRIPTION
**Pull request by sanjsingh07**
Problem
inner_call() function for hashing syscalls (sha256, keccak and blake3) are nearly identical

Summary of Changes
de-replicated the hashing syscalls to follow generic and trait approach, ref https://github.com/solana-labs/solana/issues/33046.

image

### CODE REVIEW

**Review by dmakarov — APPROVED** [review]
LGTM

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (t-nelson) [inline programs/bpf_loader/src/syscalls/mod.rs:1787]: we can't just use the sha256 compute budget values for all digests.  they should be queried with another method on the trait.

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (t-nelson) [inline programs/bpf_loader/src/syscalls/mod.rs:1790]: this omits the digest name from the log message.  add a `const NAME: &'static str` associated constant to the trait and set it for each

**Review by t-nelson — CHANGES_REQUESTED** [review]

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (sanjsingh07) [reply programs/bpf_loader/src/syscalls/mod.rs]: Done!

Comment (sanjsingh07) [reply programs/bpf_loader/src/syscalls/mod.rs]: so I went with the default values of sha256 in compute_budget variables for the other hashes as well since I didn't know which values they are supposed to use as previously they were using the same ones as sha256 so if they need to be updated lmk.

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (t-nelson) [inline programs/bpf_loader/src/syscalls/mod.rs:167]: instead of abusing the compute budget here, let's add a method for each arg that _takes_ a `&ComputeBudget` and return the corresponding value
```
fn get_base_cost(compute_budget: &ComputeBudget) -> u64;
fn get_byte_cost(compute_budget: &ComputeBudget) -> u64;
fn get_max_slices() -> u64;
```

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (t-nelson) [inline programs/bpf_loader/src/syscalls/mod.rs:141]:
```suggestion
    fn result(self) -> [u8; Self::HASH_BYTES];
```
(and updates throughout)

otherwise i believe the outter scoped `hash::HASH_BYTES` is in scope

**Review by t-nelson — COMMENTED** [review]
now we're headed in the right direction! thanks for the changes

---

> so I went with the default values of sha256 in compute_budget variables for the other hashes as well since I didn't know which values they are supposed to use as previously they were using the same ones as sha256 so if they need to be updated lmk.

hrm... wouldn't have expected that. we should definitely properly measure them, but that's out of scope fore this change set. i'd expect at leaset blake3 to be cheaper than sha256

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (sanjsingh07) [reply programs/bpf_loader/src/syscalls/mod.rs:141]: so I tried updating it but then started getting this error "generic parameters may not be used in const operations" and "type parameters may not be used in const expressions" So the only 2 solutions I could think of is either go with constant numerical value '32' or just use outer scoped hash::HASH_BYTES for trait only and for the implementations, can use 'Self::HASH_BYTES'.

File: program-runtime/src/compute_budget.rs
Comment (sanjsingh07) [reply program-runtime/src/compute_budget.rs]: reverted all the changes back to the original state.

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (sanjsingh07) [reply programs/bpf_loader/src/syscalls/mod.rs]: added these above methods and they return the corresponding value, but as mentioned in the above conversation all hashing methods use the same `base_cost`, `max_salice`, and `byte_cost` since they never had different values so they should be taken care of in another PR/issue.

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (t-nelson) [reply programs/bpf_loader/src/syscalls/mod.rs:141]: oh that's probably from the array definition :thinking: 

maybe instead of the associated constant on the trait. declare an associated _type_?

something like
```
type Output: AsRef<[u8]>;
```

`AsRef<u8>` should be enough for the `copy_from_slice()` we do to send the result back to vm.

this method then becomes
```
fn result(self) -> Self::Output;
```

Comment (sanjsingh07) [reply programs/bpf_loader/src/syscalls/mod.rs]: Done!

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (t-nelson) [inline programs/bpf_loader/src/syscalls/mod.rs:155]: we should be able to use each module's `Hash` type here instead of a raw byte array. then we don't need `HASH_BYTES` at all, instead we can replace the usage when calling `translate_slice_mut()` with `std::mem::<H::Output>::size_of()` and call `as_ref()` on it in the `copy_from_slice()` usage

**Review by t-nelson — COMMENTED** [review]
almost there!

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (sanjsingh07) [reply programs/bpf_loader/src/syscalls/mod.rs]: done!

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (t-nelson) [inline programs/bpf_loader/src/syscalls/mod.rs:1905]: i _think_ clippy's going to complain about the `&` here, which is implicit from `.as_ref()`, but let's see what it says before changing

**Review by t-nelson — COMMENTED** [review]
ok. looks good to me know.  let's see what the robot says!

**Comment by t-nelson** [comment]
> ok. looks good to me know. let's see what the robot says!

just realized you may not be able to view the ci output... looks like it's not liking trailing whitespace at `programs/bpf_loader/src/syscalls/mod.rs:232`. if you want to iterate faster, get `./ci/test-sanity.sh` and `./ci/test-checks.sh` passing locally. those are typically the bulk of the issues with refactors like this and pretty easy to run locally.

**Comment by sanjsingh07** [comment]
> > ok. looks good to me know. let's see what the robot says!
> 
> just realized you may not be able to view the ci output... looks like it's not liking trailing whitespace at `programs/bpf_loader/src/syscalls/mod.rs:232`. if you want to iterate faster, get `./ci/test-sanity.sh` and `./ci/test-checks.sh` passing locally. those are typically the bulk of the issues with refactors like this and pretty easy to run locally.

Hey, I was able to see the ci output, and I ran `cargo fmt -all` to deal with the white space issue, `cargo clippy` to look for unused and reference related issue, which I've fixed/updated in new commits and also ` ./ci/test-sanity.sh` ran completely fine but `./ci/test-checks.sh` failed with 

```
+ exec cargo +nightly-2023-08-25 check --version --verbose
error: unexpected argument '--version' found

  tip: a similar argument exists: '--verbose'

Usage: cargo check --verbose...

For more information, try '--help'.
```

**Comment by t-nelson** [comment]
> > > ok. looks good to me know. let's see what the robot says!
> > 
> > 
> > just realized you may not be able to view the ci output... looks like it's not liking trailing whitespace at `programs/bpf_loader/src/syscalls/mod.rs:232`. if you want to iterate faster, get `./ci/test-sanity.sh` and `./ci/test-checks.sh` passing locally. those are typically the bulk of the issues with refactors like this and pretty easy to run locally.
> 
> Hey, I was able to see the ci output, and I ran `cargo fmt -all` to deal with the white space issue, `cargo clippy` to look for unused and reference related issue, which I've fixed/updated in new commits and also ` ./ci/test-sanity.sh` ran completely fine but `./ci/test-checks.sh` failed with below code snippet, and I tried I updating the `hack` sub-cmd with `check` which led to different error so you also might wanna check that out
> 
> ```
> + exec cargo +nightly-2023-08-25 hack --version --verbose
> error: no such command: `hack`
> 
>         Did you mean `check`?
> 
>         View all installed commands with `cargo --list`
> ```

ah you need to `cargo install cargo-hack`, i believe

**Comment by t-nelson** [comment]
@sanjsingh07 you don't need to worry about the "this branch is outdated..." warning, so long as it doesn't escalate to "this branch conflicts...". rebasing clears ci, so we'll need to rerun to merge

@Lichtso @alessandrod can i get one of you to sign off on this again?  it's ready to merge otherwise

**Comment by sanjsingh07** [comment]
> @sanjsingh07 you don't need to worry about the "this branch is outdated..." warning, so long as it doesn't escalate to "this branch conflicts...". rebasing clears ci, so we'll need to rerun to merge
> 
> @Lichtso @alessandrod can I get one of you to sign off on this again? it's ready to merge otherwise

Oh ok, Thanks for letting me know about that cuz I wasn't sure if it was mergeable with the branch being outdated.

**Comment by t-nelson** [comment]
> Oh ok, Thanks for letting me know about that cuz I wasn't sure if it was mergeable with the branch being outdated.

no worries, github doesn't exactly make it obvious!

once we get another sign-off from the runtime team, this will be ready to go

**Comment by t-nelson** [comment]
> is it necessary to get a sign-off from the runtime team only?

once runtime signs off this is good to merge. that team is on a few high priority tasks atm. they should be freeing up in the next few days. thanks for you patience

**Comment by sanjsingh07** [comment]
Hi @t-nelson, there is a branch conflict so should I fix and commit the issue or you can take care of it when merging?

**Comment by sanjsingh07** [comment]
> Hi @t-nelson, there is a branch conflict so should I fix and commit the issue or you can take care of it when merging?

nevermind, had free time so did it myself.

**Comment by t-nelson** [comment]
@Lichtso @alessandrod @dmakarov can we get a final signoff from runtime on this? should be net better than the last time it was approved 😉

**Comment by t-nelson** [comment]
@sanjsingh07 looks like we picked up some trailing whitespace somewhere. ci sanity tests are failing

**Comment by sanjsingh07** [comment]
> @sanjsingh07 looks like we picked up some trailing whitespace somewhere. ci sanity tests are failing

yeah my bad, I might have left that while resolving the branch conflict, so do you want me to push the updated code or you can take care of it while merging?

**Review by alessandrod — COMMENTED** [review]
lgtm! Please remove the whitespace at line 1951 then it's good to go

**Comment by sanjsingh07** [comment]
> lgtm! Please remove the whitespace at line 1951 then it's good to go

done!

**Review by t-nelson — APPROVED** [review]
we're r+ ci here. merge when green

**Comment by t-nelson** [comment]
finally made it!  thanks for the contribution and patience with review responses!

**Pull request merged**
